Initial documentation for GitHub dependency graph integration (#86)
microsoft/vcpkg-docs: vcpkg/users/github-dependency-graph.md

---
title: Integrating with GitHub dependency graph
description: Record the set of resolved libraries with the GitHub dependency graph
ms.date: 05/22/2023
---

# Integrating with the GitHub dependency graph

> 

Reviewer: Needs to add the experimental block, see how it is done for the asset caching docs

> [!TIP]

> > This page describes an experimental feature of vcpkg. We are actively developing this feature and would like to hear your feedback. Let us know if you have any thoughts about the current functionality or any desired functionality by emailing vcpkg@microsoft.com or by filing an issue in [vcpkg-docs](https://github.com/microsoft/vcpkg-docs/issues). If you have feedback on any of the GitHub features that this enables, let us know tha too and we'll make sure it gets to the right folks at GitHub.

Author: Make email address clickable.

## About the GitHub dependency graph

The GitHub [dependency graph](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-the-dependency-graph) stores the set of dependencies for a repository. Beyond just being able to visualize what a repository's dependencies are, GitHub builds several useful features on top of this data, including [dependency review](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-dependency-review) and [Dependabot alerts](https://docs.github.com/en/code-security/dependabot/dependabot-alerts/about-dependabot-alerts).

> ## vcpkg integration with the GitHub depdency graph

Reviewer: typo `dependency`

vcpkg has experimental support for populating the GitHub dependency graph from within a GitHub Action workflow. In order to enable this feature, you must include the following in your workflow file:

* Set the `dependencygraph` feature flag to the environment variable `VCPKG_FEATURE_FLAGS`.

> * Set the `GITHUB_TOKEN` environment variable with the GitHub workflow variable `${{ secrets.GITHUB_TOKEN }}`.

Author: In retrospect, the wording here feels weird. Open to suggestions, particularly if vcpkg docs have a standard way of phrasing this type of stuff.

> * Enable write access to the contents of your repository by including the following permissions block. This is because it must write dependency graph metadata to your repository. This __will not__ add any commits to your repository or otherwise modify the contents of your repository (despite the permission name).

Reviewer: Perhaps we should add a section here asking users to enable dependency graph support for their repo

Author: I think that's a good idea. If it's a public repository, it's enabled by default, but if it's not, then we should give them an idea of how to do it. Maybe linking to some GitHub doc would be sufficient. I'll see what makes most sense (I know I saw somewhere someone trying to invent a button that could easily do it, but I think the GH settings UI has improved since that time).

```yaml
permissions:
  contents: write
```

## Known limitations

The following are known limitations in the current implementation. __Even though we know of these, if they affect the value for you, we'd still like to hear from you so that we can prioritize what we need to fix and how we should fix it.__

* The version of vcpkg that is bundled with GitHub Actions runners does not have this feature yet. You must use invoke the correct version of vcpkg that supports this feature

> * TODO

Author: Need to make clear that the dependency graph won't enable some of the downstream GitHub features on it's own. We still want to hear from people that want those features though if they really want them.

Author: The latest update should have added that and removed this TODO line.

## Example GitHub Actions workflow

> 

Author: Hm. I'm not too crazy about tell them what they should name it. The main reason why is that for people we are looking to adopt, they likely already have a workflow YAML file declared, but since this is an example, maybe it's okay. I just know that I would instantly start questioning whether it needs to be named `pipeline.yaml`.

```yaml
name: Populate dependencies

on:
  push:
    branches: [ main ]
  workflow_dispatch:

permissions:
  contents: write

env:
  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
  VCPKG_FEATURE_FLAGS: dependencygraph

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v3
      with:
        submodules: 'recursive'

    # This will execute a dry-run, meaning that libraries will not be built and
    # installed, but they will still be reported to the GitHub dependency graph.
    # This step assumes `vcpkg` has been bootstrapped (run `./vcpkg/bootstrap-vcpkg`)
    - name: Run vcpkg

>       run: ${{ github.workspace }}/vcpkg/vcpkg install --dry-run

Author: There should be a manifest. I think that this would be a good place to consider the complete example I have in the template repository. I could link there instead. I wonder if there's a way to include the contents of a remote file like that here in Learn docs.

>       run: ${{ github.workspace }}/vcpkg/vcpkg install --dry-run

Author: I'll put a comment on the example noting the assumption that there is a valid `vcpkg.json` manifest in place.

---
title: Integrating with GitHub dependency graph
description: Record the set of resolved libraries with the GitHub dependency graph
ms.date: 05/22/2023
---

# Integrating with the GitHub dependency graph

[!INCLUDE [experimental](../../includes/experimental.md)]

> [!TIP]
> We are actively developing this feature and would like to hear your feedback. Let us know if you have any thoughts about the current functionality or any desired functionality by emailing [vcpkg@microsoft.com](mailto:vcpkg@microsoft.com) or by filing an issue in [vcpkg-docs](https://github.com/microsoft/vcpkg-docs/issues). If you have feedback on any of the GitHub features that this enables, let us know that too and we'll make sure it gets to the right folks at GitHub.

## About the GitHub dependency graph

> The GitHub [dependency graph](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-the-dependency-graph) stores the set of dependencies for a repository. Beyond just being able to visualize what a repository's dependencies are, GitHub builds several useful features on top of this data, including [dependency review](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-dependency-review) and [Dependabot alerts](https://docs.github.com/en/code-security/dependabot/dependabot-alerts/about-dependabot-alerts).

Reviewer 2: ```suggestion
The GitHub [dependency graph](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-the-dependency-graph) stores the set of dependencies for a repository. Beyond being able to visualize a repository's dependencies, GitHub builds several useful features on top of this data, including [dependency review](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-dependency-review) and [Dependabot alerts](https://docs.github.com/en/code-security/dependabot/dependabot-alerts/about-dependabot-alerts).
```

## vcpkg integration with the GitHub dependency graph

> vcpkg has experimental support for populating the GitHub dependency graph from within a GitHub Action workflow. In order to enable this feature, you must include the following in your workflow file:

Reviewer 2: ```suggestion
vcpkg has experimental support for populating the GitHub dependency graph from within a GitHub Actions workflow. To enable this feature, make the following changes in your workflow file:
```

* Set the `dependencygraph` feature flag to the environment variable `VCPKG_FEATURE_FLAGS`.
* Set the `GITHUB_TOKEN` environment variable with the GitHub workflow variable `${{ secrets.GITHUB_TOKEN }}`.

> * Enable write access to the contents of your repository by including the following permissions block. This is because it must write dependency graph metadata to your repository. This __will not__ add any commits to your repository or otherwise modify the contents of your repository (despite the permission name).

Reviewer 2: ```suggestion
* Give GitHub Actions write access to the contents of your repository. This is required to write the dependency graph metadata into your repository. This __will not__ add any commits to your repository or make any other modification to the contents of your repository.
```

```yaml
permissions:
  contents: write
```

> Additionally, you must have the GitHub dependency graph enabled. If your repository is a public repository, it is enabled by default. If it is a private repository, follow the [GitHub documentation to enable the dependency graph](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/configuring-the-dependency-graph#enabling-and-disabling-the-dependency-graph-for-a-private-repository) in order to satisfy this requirement.

Reviewer 2: ```suggestion
Additionally, you must enable the GitHub dependency graph in your repository's settings (enabled by default on public repositories). On private repositories, follow the [GitHub documentation to enable the dependency graph](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/configuring-the-dependency-graph#enabling-and-disabling-the-dependency-graph-for-a-private-repository).
```

## Known limitations

The following are known limitations in the current implementation. __Even though we know of these, if they affect the value for you, we'd still like to hear from you so that we can prioritize what we need to fix and how we should fix it.__

> * The version of vcpkg that is bundled with GitHub Actions runners does not have this feature yet. You must use invoke the correct version of vcpkg that supports this feature

Reviewer 2: ```suggestion
* The version of vcpkg that is bundled with GitHub Actions runners does not have this feature yet. Ensure you're using a version of vcpkg that supports this feature.
```

> * Features that depend on the dependency graph, such as Dependbot alerts and Dependabot pull requests, are not yet available. Please let us know if you are interested in those features!

Reviewer 2: ```suggestion
* Features that depend on the dependency graph, such as Dependabot alerts and Dependabot pull requests, are not yet available. Please let us know if you are interested in those features!
```
